backend/store/memory/event_store.go

```diff
@@ -0,0 +1,366 @@
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	corev2 "github.com/sensu/core/v2"
+	"github.com/sensu/sensu-go/backend/store"
+	"golang.org/x/time/rate"
+)
+
+type memorydb struct {
+	limiter *rate.Limiter
+	data    sync.Map
+}
+
+func newMemoryDB(limiter *rate.Limiter) *memorydb {
+	return &memorydb{
+		limiter: limiter,
+	}
+}
+
+type eventEntry struct {
+	// Serialized and compressed event
+	EventBytes []byte
+
+	// Dirty indicates if the entry has yet to be written out to long term storage
+	Dirty bool
+
+	// Mu is necessary in the rare instance when multiple goroutines are competing
+	// to store different versions of the same event. This is only necessary due to
+	// a design flaw in eventd, and could possibly be mitigated there.
+	Mu sync.Mutex
+}
+
+const eventOutputTruncatedBytesLabel = "sensu.io/output_truncated_bytes"
+
+type SilenceStoreI interface {
+	// DeleteSilences deletes all silences matching the given names.
+	DeleteSilences(ctx context.Context, namespace string, names []string) error
+
+	// GetSilences returns all silences in the namespace. A nil slice with no error is
+	// returned if none were found.
+	GetSilences(ctx context.Context, namespace string) ([]*corev2.Silenced, error)
+
+	// GetSilencedsByCheck returns all silences for the given check . A nil
+	// slice with no error is returned if none were found.
+	GetSilencesByCheck(ctx context.Context, namespace, check string) ([]*corev2.Silenced, error)
+
+	// GetSilencedEntriesBySubscription returns all entries for the given
+	// subscription. A nil slice with no error is returned if none were found.
+	GetSilencesBySubscription(ctx context.Context, namespace string, subscriptions []string) ([]*corev2.Silenced, error)
+
+	// GetSilenceByName returns an entry using the given namespace and name. An
+	// error is returned if the entry is not found.
+	GetSilenceByName(ctx context.Context, namespace, name string) (*corev2.Silenced, error)
+
+	// UpdateSilence creates or updates a given silence.
+	UpdateSilence(ctx context.Context, silence *corev2.Silenced) error
+
+	// GetSilencesByName gets all the named silence entries.
+	GetSilencesByName(ctx context.Context, namespace string, names []string) ([]*corev2.Silenced, error)
+}
+
+func (m *memorydb) ReadEntry(namespace, entity, check string) *eventEntry {
+	key := strings.Join([]string{namespace, entity, check}, "\n")
+	emptyEntry := &eventEntry{}
+	result, _ := m.data.LoadOrStore(key, emptyEntry)
+	entry := result.(*eventEntry)
+	return entry
+}
+
+func (m *memorydb) WriteTo(ctx context.Context, s store.EventStore) error {
+	var storeErr error
+	ctx = store.NoMergeEventContext(ctx)
+	m.data.Range(func(key, value any) bool {
+		entry := value.(*eventEntry)
+		entry.Mu.Lock()
+		defer entry.Mu.Unlock()
+		if !entry.Dirty {
+			return true
+		}
+		eventBytes, err := snappy.Decode(nil, entry.EventBytes)
+		if err != nil {
+			// developer error, fatal
+			panic(err)
```

Comment:
Is it really a developer error?

Can't `snappy.Decode()` above fail even if your code were 100% correct, say because the memory holding `entry.EventBytes` got corrupted somehow (malicious or accidental memory manipulation by something else on the system, cosmic ray flipping a bit, ...)?

Reply:
I suppose so, but that would necessitate a panic too.

Reply:
Why? Wouldn't logging the issue and skipping over the entry be a way to go about it?

Reply:
If the system has corrupted memory then it is working in an undefined state and should probably crash.